[BugFix] Avoid being unable to select when partition_duration is set incorrectly

stonechen111 opened the pull request:
Why I'm doing:
Before this submission, when partition_duration was set incorrectly, selection
could not be performed and "Unsupported unit: Seconds" was returned directly.

What I'm doing:
This submission modifies this behavior: when partition_duration is set incorrectly,
we convert the datacache.partition_duration type from date to datetime to compare
with the partition range and catching the other exceptions (if any) during time comparison.

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

imay commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -2984,20 +2985,24 @@ private boolean isEnableFillDataCacheImpl(Partition partition) throws AnalysisEx
+                    // enable data cache by default.
+                    return true;
+                }
+            } catch (UnsupportedTemporalTypeException e) {
+                LOG.error(e.getMessage());
```
1. you need to comment when this exception will be thrown.
2. change LOG.error to LOG.warn, error often be used for system errors.

stonechen111 replied:
OK

kevincai commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -2990,10 +2991,15 @@ private boolean isEnableFillDataCacheImpl(Partition partition) throws AnalysisEx
+                    dataCacheRange = Range.openClosed(PartitionKey.ofDate(lower), PartitionKey.ofDate(upper));
+                    return partitionRange.isConnected(dataCacheRange);
+                } catch (UnsupportedTemporalTypeException e) {
+                    LOG.warn("{}. table name: {}, datacache.partition_duration: {}, mismatch with partition range.",
```
so it will return `true` anyway?

need add UT to cover this `isEnableFillDataCache()` interface.

stonechen111 committed:
[BugFix] Avoid being unable to select when partition_duration is set incorrectly

Why I'm doing:

Before this submission, when partition_duration was set incorrectly, selection
could not be performed and "Unsupported unit: Seconds" was returned directly.

What I'm doing:

This submission modifies this behavior: when partition_duration is set incorrectly,
we convert the datacache.partition_duration type from date to datetime to compare
with the partition range and catching the other exceptions (if any) during time comparison.

Signed-off-by: stonechen111 <chenshi@starrocks.com>

stonechen111 committed:
Modify test cases

Signed-off-by: stonechen111 <chenshi@starrocks.com>
